Mobile/ShopLens/Droid/Fragments/Camera2Fragment.cs

```diff
@@ -0,0 +1,550 @@
+                {
+                    mCaptureSession.Close();
+                    mCaptureSession = null;
+                }
+                if (null != mCameraDevice)
+                {
+                    mCameraDevice.Close();
+                    mCameraDevice = null;
+                }
+                if (null != mImageReader)
+                {
+                    mImageReader.Close();
+                    mImageReader = null;
+                }
+            }
+            catch (InterruptedException e)
+            {
+                throw new RuntimeException("Interrupted while trying to lock camera closing.", e);
+            }
+            finally
+            {
+                mCameraOpenCloseLock.Release();
+            }
+        }
+
+        // Starts a background thread and its {@link Handler}.
+        private void StartBackgroundThread()
+        {
+            mBackgroundThread = new HandlerThread("CameraBackground");
+            mBackgroundThread.Start();
+            mBackgroundHandler = new Handler(mBackgroundThread.Looper);
+        }
+
+        // Stops the background thread and its {@link Handler}.
+        private void StopBackgroundThread()
+        {
+            mBackgroundThread.QuitSafely();
+            try
+            {
+                mBackgroundThread.Join();
+                mBackgroundThread = null;
+                mBackgroundHandler = null;
+            }
+            catch (InterruptedException e)
+            {
+                e.PrintStackTrace();
+            }
+        }
+
+        // Creates a new {@link CameraCaptureSession} for camera preview.
+        public void CreateCameraPreviewSession()
+        {
+            try
+            {
+                SurfaceTexture texture = mTextureView.SurfaceTexture;
+                if (texture == null)
+                {
+                    throw new IllegalStateException("texture is null");
+                }
+
+                // We configure the size of default buffer to be the size of camera preview we want.
+                texture.SetDefaultBufferSize(mPreviewSize.Width, mPreviewSize.Height);
+
+                // This is the output Surface we need to start preview.
+                Surface surface = new Surface(texture);
+
+                // We set up a CaptureRequest.Builder with the output Surface.
+                mPreviewRequestBuilder = mCameraDevice.CreateCaptureRequest(CameraTemplate.Preview);
+                mPreviewRequestBuilder.AddTarget(surface);
+
+                // Here, we create a CameraCaptureSession for camera preview.
+                List<Surface> surfaces = new List<Surface>
+                {
+                    surface,
+                    mImageReader.Surface
+                };
+                mCameraDevice.CreateCaptureSession(surfaces, new CaptureSessionCallback(this), null);
+
+            }
+            catch (CameraAccessException e)
+            {
+                e.PrintStackTrace();
```

Comment:
I think you should make your custom exception for CameraAccessException, because this exception is catched so many times.